File: src/java/eu/sig/training/ch02/v1/BoardFactory.java
package eu.sig.training.ch02.v1;

public class BoardFactory {
    // tag::createBoard[]
    public Board createBoard(Square[][] grid) {
        assert grid != null;

        Board board = new Board(grid);

        int width = board.getWidth();
        int height = board.getHeight();
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                Square square = grid[x][y];
                for (Direction dir : Direction.values()) {
                    setLink(square, dir, x, y, width, height, grid);
                }
            }
        }

        return board;
    }
    // end::createBoard[]

    // tag::setLink[]
    private void setLink(Square square, Direction dir, int x, int y, int width,
        int height, Square[][] grid) {
        int dirX = (width + x + dir.getDeltaX()) % width;
        int dirY = (height + y + dir.getDeltaY()) % height;
        Square neighbour = grid[dirX][dirY];
        square.link(neighbour, dir);
    }
    // end::setLink[]
}

class Board {
    @SuppressWarnings("unused")
    public Board(Square[][] grid) {}

    public int getWidth() {
        return 0;
    }

    public int getHeight() {
        return 0;
    }
}

class Square {
    @SuppressWarnings("unused")
    public void link(Square neighbour, Direction dir) {}
}

class Direction {

    public static Direction[] values() {
        return null;
    }

    public int getDeltaY() {
        return 0;
    }

    public int getDeltaX() {
        return 0;
    }
}

File: src/csharp/eu/sig/training/ch02/v1/BoardFactory.cs
using System.Diagnostics;

namespace eu.sig.training.ch02.v1
{
    public class BoardFactory
    {
        // tag::createBoard[]
        public Board CreateBoard(Square[,] grid)
        {
            Debug.Assert(grid != null);

            Board board = new Board(grid);

            int width = board.Width;
            int height = board.Height;
            for (int x = 0; x < width; x++)
            {
                for (int y = 0; y < height; y++)
                {
                    Square square = grid[x, y];
                    foreach (Direction dir in Direction.Values)
                    {
                        SetLink(square, dir, x, y, width, height, grid);
                    }
                }
            }

            return board;
        }
        // end::createBoard[]

        // tag::setLink[]
        private void SetLink(Square square, Direction dir, int x, int y,
            int width, int height, Square[,] grid)
        {
            int dirX = (width + x + dir.DeltaX) % width;
            int dirY = (height + y + dir.DeltaY) % height;
            Square neighbour = grid[dirX, dirY];
            square.Link(neighbour, dir);
        }
        // end::setLink[]
    }

    public class Board
    {
        public Board(Square[,] grid) { }

        public int Width { get; set; }
        public int Height { get; set; }
    }

    public class Square
    {
        public void Link(Square neighbour, Direction dir) { }
    }

    public class Direction
    {

        public static Direction[] Values { get; set; }

        public int DeltaY { get; set; }

        public int DeltaX { get; set; }
    }
}
